vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSPackageRegistry.java

```diff
@@ -0,0 +1,570 @@
+                    if (dep.matches(id)) {
+                        usages.add(pid);
+                        break;
+                    }
+                }
+            }
+        }
+        return usages.toArray(new PackageId[usages.size()]);
+    }
+
+    @Nonnull
+    @Override
+    public DependencyReport analyzeDependencies(@Nonnull PackageId id, boolean onlyInstalled)
+            throws IOException, NoSuchPackageException {
+        List<Dependency> unresolved = new LinkedList<Dependency>();
+        List<PackageId> resolved = new LinkedList<PackageId>();
+        try (RegisteredPackage pkg = open(id)) {
+            if (pkg == null) {
+                throw new NoSuchPackageException().setId(id);
+            }
+            // noinspection resource
+            for (Dependency dep : pkg.getPackage().getDependencies()) {
+                PackageId resolvedId = resolve(dep, onlyInstalled);
+                if (resolvedId == null) {
+                    unresolved.add(dep);
+                } else {
+                    resolved.add(resolvedId);
+                }
+            }
+        }
+
+        return new DependencyReportImpl(id, unresolved.toArray(new Dependency[unresolved.size()]),
+                resolved.toArray(new PackageId[resolved.size()]));
+    }
+
+    @Nonnull
+    @Override
+    public PackageId register(@Nonnull InputStream in, boolean replace) throws IOException, PackageExistsException {
+        ZipVaultPackage pkg = upload(in, replace);
+        return pkg.getId();
+    }
+
+    public ZipVaultPackage upload(InputStream in, boolean replace)
+            throws IOException, PackageExistsException {
+
+        MemoryArchive archive = new MemoryArchive(true);
+        File tempFile = File.createTempFile("upload", ".zip");
+        
+        InputStreamPump pump = new InputStreamPump(in , archive);
+        try {
+            // this will cause the input stream to be consumed and the memory
+            // archive being initialized.
+            try {
+
+                FileUtils.copyInputStreamToFile(pump, tempFile);
+            } catch (Exception e) {
+                String msg = "Stream could be read successfully.";
+                log.error(msg);
+                throw new IOException(msg, e);
+            }
+    
+            if (archive.getJcrRoot() == null) {
+                String msg = "Stream is not a content package. Missing 'jcr_root'.";
+                log.error(msg);
+                throw new IOException(msg);
+            }
+    
+            final MetaInf inf = archive.getMetaInf();
+            PackagePropertiesImpl props = new PackagePropertiesImpl() {
+                @Override
+                protected Properties getPropertiesMap() {
+                    return inf.getProperties();
+                }
+            };
+            PackageId pid = props.getId();
+    
+            // invalidate pid if path is unknown
+            if (pid == null) {
```

Comment:
this is rather for backwards compatibility. I wonder if we should still allow this...

Reply:
Not sure.. I think we could also throw an IOException and no longer support it.